### PR TITLE
Script for calculating the deviating between caffe output model and onnx output model

### PR DESCRIPTION
**Pull request by vodanhbk95**
**Update code**  in 27/07/2020

In this branch, I add the following things:

- Add the script to calculate the MSE between the caffe model and  onnx model [verify_caffe_model.py](https://github.com/cybercore-co-ltd/Onnx2Caffe/blob/compare_weight_pytorch_caffe/verify_caffe_model.py)
- Remove **test_img.jpg**

## Experiments
- With the random the input_data, I calculate the mae_error and rel_err through 5 stages (`rel_err` was computed by `mae/norm`)

| Stage | Branch       | MAE           | Relative_err   |
|-------|--------------|---------------|----------------|
| 1     | cls_score.0  | 0.021983385   | 3.72074e-05    |
| 2     | cls_score.1  | 0.0049471855  | 1.42990075e-05 |
| 3     | cls_score.2  | 0.00044226646 | 2.8996503e-06  |
| 4     | cls_score.3  | 0.00010037422 | 1.4506386e-06  |
| 5     | cls_score.4  | 2.2172928e-05 | 6.535266e-07   |
| 1     | bbox_pred.0  | 0.056485035   | 0.00016337779  |
| 2     | bbox_pred.1  | 0.021685213   | 0.00013785706  |
| 3     | bbox_pred.2  | 0.003439759   | 3.7137896e-05  |
| 4     | bbox_pred.3  | 0.00046535768 | 1.5086444e-05  |
| 5     | bbox_pred.4  | 8.970499e-05  | 4.651793e-06   |
| 1     | centerness.0 | 0.005687684   | 4.833609e-05   |
| 2     | centerness.1 | 0.0011602044  | 2.5652998e-05  |
| 3     | centerness.2 | 0.00025871396 | 1.472214e-05   |
| 4     | centerness.3 | 4.413724e-05  | 4.627184e-06   |
| 5     | centerness.4 | 9.298325e-06  | 9.152293e-06   |

## Result
TOTAL ERR BETWEEN CAFFE MODEL AND ONNX MODEL : TOTAL_ERR 0.0005171119058786644

## Command test
```
python verify_caffe_model.py <onnx_path>  <caffe_path> <prototxt_path>
```

### CODE REVIEW

File: tools/verify_caffe_model.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    for num in range(len(num_layer)):
+
+        # calculate the mae error between onnx and caffe model
+        mae_err = np.abs(caffe_outs[num_layer[num]] - onnx_result[num]).sum()
+        
+        # calculate the relative err mae/norm(oonx)
+        norm_onnx = np.linalg.norm(onnx_result[num])
+        rel_err = mae_err / norm_onnx
+        total_mae_err += mae_err
+        total_rel_err += rel_err
```
Comment (chuong98) [inline tools/verify_caffe_model.py:80]: Please use terminaltable to print the error at each layer. Check the same function in 
https://github.com/cybercore-co-ltd/ccdetection/blob/master/tools/verify_onnx_model.py

And then copy the results table, and show it in your report

File: tools/verify_caffe_model.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    input_name = str(graph.inputs[0][0])
+    caffe_model.blobs[input_name].data[...] = inputs
+    caffe_outs = caffe_model.forward()
+    
+    return caffe_outs
+
+def compute_relative_err_onnx2caffe(onnx_result, caffe_outs):
+
+    num_layer = ['490', '510', '530', '550', '570', '492', '512',
```
[the remark this answers is not in the document]
Comment (vodanhbk95) [reply tools/verify_caffe_model.py:67]: Many thanks. I got it & It did well. I have waste much time in this step.

File: tools/verify_caffe_model.py [hunk 3]
```diff
@@ -8,8 +8,6 @@
 import onnxruntime as rt
 import torch
 
-from mmcv.runner import load_checkpoint
-from ccdet.models import build_detector
 from convertCaffe import convertToCaffe, getGraph
 from terminaltables import AsciiTable
 from ccdet.utils import PublicConfig as Config
```
Comment (chuong98) [inline tools/verify_caffe_model.py]: NOO, Why still have this one. 
`from ccdet.utils import PublicConfig as Config`

+ I don't want to see any thing related to `ccdet` or `mmdet`
+ I don't want to see any python config in here.
The ONNX file has everything you need to convert to Caffe, include:
+ output name:
```python
model = onnx.load('xxx.onnx')
 output_name =[node.name for node in model.graph.output]
```
+ Input_name, input_shape: 
https://stackoverflow.com/questions/56734576/find-input-shape-from-onnx-file

File: verify_caffe_model.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+            # table result
+            table_data.append([j+1, output_names[5*i+j], mae_err, rel_err])
+
+    table = AsciiTable(table_data)
+    print(table.table)
+
+    return total_err
+
+def auto_rename_outputs(cfg):
```
Comment (chuong98) [inline verify_caffe_model.py:111]: NOOOOO, this is the same of hard coding the name. 
Please delete the function `auto_rename_ouputs`
Use this to get the name.
```python
model = onnx.load('xxx.onnx')
output_name =[node.name for node in model.graph.output]
```

File: verify_caffe_model.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    elif len(args.shape) == 2:
+        input_shape = (
+            1,
+            3,
+        ) + tuple(args.shape)
+    else:
+        raise ValueError('invalid input shape')
+    
+    # parse the config
+    cfg = Config.fromfile(args.config)
```
Comment (chuong98) [inline verify_caffe_model.py:138]: Remove this. I don't want to see config file.

File: verify_caffe_model.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+    # get name of output features: cls branch, bbox_branch, centerness_branch
+    output_names = auto_rename_outputs(cfg)
```
Comment (chuong98) [inline verify_caffe_model.py:141]: Delete this. Replace by a function: `get_onnx_outputname`

**Comment by chuong98** [comment]
+ Onnx2caffe is separate package. It should work in general, without `ccdet` or `mmdet `.
 + Onnx model itself has all information of the NW. Don't ever again use config_file as the input.

File: tools/verify_caffe_model.py (hunk 3 above)
Comment (vodanhbk95) [reply tools/verify_caffe_model.py]: If I use 
```
 model = onnx.load('xxx.onnx')
 output_name =[node.name for node in model.graph.output]
```
It returns output_name : ['490', '510', '530', '550', '570', '492', '512', '532', '552', '572', '493', '513', '533', '553', '573']
And I'm not sure anybody understand this node is based on which branch. That is the reason why I use **"config model"** to show clearly the name of output in the branch.

Comment (chuong98) [reply tools/verify_caffe_model.py]: Did you use correct ONNX file? THe Onnx file must be exported with "auto_rename". Check the tutorial of Export Onnx

File: verify_caffe_model.py (hunk 4 above)
Comment (vodanhbk95) [reply verify_caffe_model.py:111]: Done

File: verify_caffe_model.py (hunk 5 above)
Comment (vodanhbk95) [reply verify_caffe_model.py:138]: Done

File: verify_caffe_model.py (hunk 6 above)
Comment (vodanhbk95) [reply verify_caffe_model.py:141]: Done

File: tools/verify_caffe_model.py (hunk 3 above)
Comment (vodanhbk95) [reply tools/verify_caffe_model.py]: Done

File: verify_caffe_model.py [hunk 7]
```diff
@@ -82,44 +63,37 @@ def get_caffe_pred(model, inputs):
     
     return caffe_outs
 
-def compute_relative_err_onnx2caffe(onnx_result, caffe_outs, output_names):
+def compute_relative_err_onnx2caffe(onnx_result, caffe_outs, output_name):
 
     total_err = 0
     table_data = [
         ['Stage', 'Branch', 'MAE', 'Relative_err']
```
[the remark this answers is not in the document]
Comment (vodanhbk95) [reply verify_caffe_model.py]: Done.

**Pull request merged**